app/plugin.go

```diff
@@ -647,3 +426,62 @@ func (a *App) ShutDownPlugins() {
 	a.PluginConfigListenerId = ""
 	a.PluginEnv = nil
 }
+
+func (a *App) SetPluginKey(pluginId string, key string, value interface{}) *model.AppError {
+	jsonBytes, err := json.Marshal(value)
+	if err != nil {
+		return model.NewAppError("SetPluginKey", "app.plugin.key_value.set.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	kv := &model.PluginKeyValue{
+		PluginId: pluginId,
+		Key:      key,
+		Value:    string(jsonBytes),
+	}
+
+	tries := 0
+	result := store.StoreResult{}
+
+	// Retry once to deal with unique constraint errors from races in PostgreSQL query
```

Comment:
This is interesting. I can't recall seeing a retry loop like this elsewhere in the code. What constraint errors/races are you concerned about here?

Reply:
The Postgres query that CB commented on above is a bit racy since it's really two queries (an insert and an update) trying to mimic an upsert. One variation of the race is basically two different calls to the API happen at the same time to a key that doesn't exist, both for insert/update. If one call tries to updates and fails it will try to insert but if the other call inserts between the update and insert of the first call we could get a unique key violation. There's not much of a way around that by using a different query unless we're on Postgres 9.5+ but the odds are if it happens once, a quick retry will be successful

This will become unneeded as I'm going to use CB's suggestion to just use two separate queries and assume success on unique key violations (since it's a valid race)